In [1]:
import os
import requests
from dotenv import load_dotenv
import pandas as pd

pd.set_option("display.max_columns", None)

load_dotenv()

SCORECARD_KEY = os.getenv("COLLEGE_SCORECARD_API_KEY")

In [2]:
BASE_URL = "https://api.data.gov/ed/collegescorecard/v1/schools"

Utilize the CSC data dictionary (Misc/CollegeScorecardDataDictionary.xlsx) to find variable names

In [9]:
fields = ",".join([
    "school.id",
    "school.name",
    "school.ownership",
    "school.main_campus",
    "programs.cip_4_digit.code",
    "programs.cip_4_digit.title",
    "programs.cip_4_digit.credential.level",
    "programs.cip_4_digit.distance",
    "latest.programs.cip_4_digit.earnings.5_yr.overall_median_earnings",
    "latest.programs.cip_4_digit.earnings.4_yr.working_not_enrolled.overall_count",
    "latest.programs.cip_4_digit.debt.staff_grad_plus.all.eval_inst.median"
])

params= {
    "api_key": SCORECARD_KEY,
    "school.state":"FL",
    "fields": fields,
    # "latest.programs.cip_4_digit.earnings.4_yr.overall_median_earnings__not":"null",
    "latest.programs.cip_4_digit.debt.staff_grad_plus.all.eval_inst.median__not":"null",
    "page":"1",
    "per_page":"20"
}

5 year meadian is not one of the features that allow feature

In [10]:
response = requests.get(
    BASE_URL,
    params = params,
)
data = response.json()
print(data)

{'metadata': {'page': 1, 'total': 217, 'per_page': 20}, 'results': [{'latest.programs.cip_4_digit': [{'code': '5106', 'title': 'Dental Support Services and Allied Professions.', 'distance': 1, 'credential': {'level': 1}, 'earnings': {'5_yr': {'overall_median_earnings': 32076}, '4_yr': {'working_not_enrolled': {'overall_count': 254}}}, 'debt': {'staff_grad_plus': {'all': {'eval_inst': {'median': 9500}}}}}, {'code': '5107', 'title': 'Health and Medical Administrative Services.', 'distance': 1, 'credential': {'level': 1}, 'earnings': {'5_yr': {'overall_median_earnings': 30683}, '4_yr': {'working_not_enrolled': {'overall_count': 41}}}, 'debt': {'staff_grad_plus': {'all': {'eval_inst': {'median': 9500}}}}}, {'code': '5108', 'title': 'Allied Health and Medical Assisting Services.', 'distance': 1, 'credential': {'level': 1}, 'earnings': {'5_yr': {'overall_median_earnings': 30099}, '4_yr': {'working_not_enrolled': {'overall_count': 216}}}, 'debt': {'staff_grad_plus': {'all': {'eval_inst': {'me

This is the code remove the json's nested format and normalize it into a df format

In [11]:
df = pd.json_normalize(
    data["results"],
    record_path=["latest.programs.cip_4_digit"],
    meta="school.name",
    errors="ignore"
)
df

code                                              title  distance  \
0    5106    Dental Support Services and Allied Professions.         1   
1    5107        Health and Medical Administrative Services.         1   
2    5108      Allied Health and Medical Assisting Services.         1   
3    5109  Allied Health Diagnostic, Intervention, and Tr...         1   
4    5109  Allied Health Diagnostic, Intervention, and Tr...         1   
..    ...                                                ...       ...   
450  4301                  Criminal Justice and Corrections.         3   
451  4400                           Human Services, General.         1   
452  5138  Registered Nursing, Nursing Administration, Nu...         1   
453  5138  Registered Nursing, Nursing Administration, Nu...         3   
454  5202  Business Administration, Management and Operat...         1   

     credential.level  earnings.5_yr.overall_median_earnings  \
0                   1                                32076.0   
1                   1                                30683.0   
2                   1                                30099.0   
3                   1                                43170.0   
4                   2                                61133.0   
..                ...                                    ...   
450                 3                                48574.0   
451                 3                                39275.0   
452                 2                                61559.0   
453                 3                                93083.0   
454                 3                                47431.0   

     earnings.4_yr.working_not_enrolled.overall_count  \
0                                               254.0   
1                                                41.0   
2                                               216.0   
3                                                73.0   
4                                                58.0   
..                                                ...   
450                                              71.0   
451                                              91.0   
452                                             143.0   
453                                              52.0   
454                                             146.0   

     debt.staff_grad_plus.all.eval_inst.median  \
0                                         9500   
1                                         9500   
2                                         9500   
3                                        15335   
4                                        20000   
..                                         ...   
450                                      13819   
451                                      13500   
452                                       6500   
453                                       7250   
454                                      10050   

                                school.name  
0    Concorde Career Institute-Jacksonville  
1    Concorde Career Institute-Jacksonville  
2    Concorde Career Institute-Jacksonville  
3    Concorde Career Institute-Jacksonville  
4    Concorde Career Institute-Jacksonville  
..                                      ...  
450              Indian River State College  
451              Indian River State College  
452              Indian River State College  
453              Indian River State College  
454              Indian River State College  

[455 rows x 8 columns]

In [23]:
df.loc[df["earnings.5_yr.overall_median_earnings"].isna()]

code                                              title  distance  \
16   5138  Registered Nursing, Nursing Administration, Nu...         0   
19   5107        Health and Medical Administrative Services.         2   
22   5109  Allied Health Diagnostic, Intervention, and Tr...         1   
24   5123        Rehabilitation and Therapeutic Professions.         0   
27   1104                       Information Science/Studies.         0   
..    ...                                                ...       ...   
440  5202  Business Administration, Management and Operat...         2   
441  5202  Business Administration, Management and Operat...         2   
442  5203                   Accounting and Related Services.         2   
443  5207     Entrepreneurial and Small Business Operations.         2   
446  1204  Cosmetology and Related Personal Grooming Serv...         1   

     credential.level  earnings.5_yr.overall_median_earnings  \
16                  2                                    NaN   
19                  5                                    NaN   
22                  5                                    NaN   
24                  6                                    NaN   
27                  2                                    NaN   
..                ...                                    ...   
440                 1                                    NaN   
441                 2                                    NaN   
442                 1                                    NaN   
443                 1                                    NaN   
446                 1                                    NaN   

     earnings.4_yr.working_not_enrolled.overall_count  \
16                                               38.0   
19                                                NaN   
22                                                NaN   
24                                                NaN   
27                                                NaN   
..                                                ...   
440                                               NaN   
441                                               NaN   
442                                               NaN   
443                                               NaN   
446                                               NaN   

     debt.staff_grad_plus.all.eval_inst.median  \
16                                       23492   
19                                       29325   
22                                      154604   
24                                      168900   
27                                       25406   
..                                         ...   
440                                      11695   
441                                      12044   
442                                      14125   
443                                      17612   
446                                       6333   

                         school.name  
16   Concorde Career Institute-Tampa  
19           AdventHealth University  
22           AdventHealth University  
24           AdventHealth University  
27   Florida Institute of Technology  
..                               ...  
440   Hillsborough Community College  
441   Hillsborough Community College  
442   Hillsborough Community College  
443   Hillsborough Community College  
446                Cortiva Institute  

[100 rows x 8 columns]

455 degrees availible. 100 have 5 year median income NaN. 345 remaining

In [ ]:
fields = ",".join([
    "school.name",

    "latest.programs.cip_4_digit.unit_id",
    "latest.programs.cip_4_digit.school.type",
    "latest.programs.cip_4_digit.school.main_campus",

    "programs.cip_4_digit.code",
    "programs.cip_4_digit.title",
    "programs.cip_4_digit.credential.level",
    "programs.cip_4_digit.distance",

    "latest.programs.cip_4_digit.earnings.4_yr.overall_median_earnings",
    "latest.programs.cip_4_digit.earnings.4_yr.working_not_enrolled.overall_count",
    "latest.programs.cip_4_digit.debt.staff_grad_plus.all.eval_inst.median"
])

params= {
    "api_key": SCORECARD_KEY,
    "school.state":"FL",
    "fields": fields,
    "latest.programs.cip_4_digit.earnings.4_yr.overall_median_earnings__not":"null",
    "latest.programs.cip_4_digit.debt.staff_grad_plus.all.eval_inst.median__not":"null",
    "page":"1",
    "per_page":"100"
}
response = requests.get(
    BASE_URL,
    params = params,
)
data2 = response.json()
df2 = pd.json_normalize(
    data2["results"],
    record_path=["latest.programs.cip_4_digit"],
    meta="school.name",
    errors="ignore"
)
df2

code                                              title  unit_id  \
0    5106    Dental Support Services and Allied Professions.   133845   
1    5107        Health and Medical Administrative Services.   133845   
2    5108      Allied Health and Medical Assisting Services.   133845   
3    5109  Allied Health Diagnostic, Intervention, and Tr...   133845   
4    5109  Allied Health Diagnostic, Intervention, and Tr...   133845   
..    ...                                                ...      ...   
347  4301                  Criminal Justice and Corrections.   134608   
348  4400                           Human Services, General.   134608   
349  5138  Registered Nursing, Nursing Administration, Nu...   134608   
350  5138  Registered Nursing, Nursing Administration, Nu...   134608   
351  5202  Business Administration, Management and Operat...   134608   

     distance          school.type  school.main_campus  credential.level  \
0           1  Private, for-profit                   1                 1   
1           1  Private, for-profit                   1                 1   
2           1  Private, for-profit                   1                 1   
3           1  Private, for-profit                   1                 1   
4           1  Private, for-profit                   1                 2   
..        ...                  ...                 ...               ...   
347         3               Public                   1                 3   
348         1               Public                   1                 3   
349         1               Public                   1                 2   
350         3               Public                   1                 3   
351         1               Public                   1                 3   

     earnings.4_yr.overall_median_earnings  \
0                                    26765   
1                                    29667   
2                                    26428   
3                                    40294   
4                                    53958   
..                                     ...   
347                                  39116   
348                                  33802   
349                                  48957   
350                                  78190   
351                                  41430   

     earnings.4_yr.working_not_enrolled.overall_count  \
0                                                 254   
1                                                  41   
2                                                 216   
3                                                  73   
4                                                  58   
..                                                ...   
347                                                71   
348                                                91   
349                                               143   
350                                                52   
351                                               146   

     debt.staff_grad_plus.all.eval_inst.median  \
0                                         9500   
1                                         9500   
2                                         9500   
3                                        15335   
4                                        20000   
..                                         ...   
347                                      13819   
348                                      13500   
349                                       6500   
350                                       7250   
351                                      10050   

                                school.name  
0    Concorde Career Institute-Jacksonville  
1    Concorde Career Institute-Jacksonville  
2    Concorde Career Institute-Jacksonville  
3    Concorde Career Institute-Jacksonville  
4    Concorde Career Institute-Jacksonville  
..                                      ...  
347              Indian River State College  
348     

{'metadata': {'page': 1, 'total': 193, 'per_page': 20},
 'results': [{'latest.programs.cip_4_digit': [{'code': '5106',
     'title': 'Dental Support Services and Allied Professions.',
     'unit_id': 133845,
     'distance': 1,
     'school': {'type': 'Private, for-profit', 'main_campus': 1},
     'credential': {'level': 1},
     'earnings': {'4_yr': {'overall_median_earnings': 26765,
       'working_not_enrolled': {'overall_count': 254}}},
     'debt': {'staff_grad_plus': {'all': {'eval_inst': {'median': 9500}}}}},
    {'code': '5107',
     'title': 'Health and Medical Administrative Services.',
     'unit_id': 133845,
     'distance': 1,
     'school': {'type': 'Private, for-profit', 'main_campus': 1},
     'credential': {'level': 1},
     'earnings': {'4_yr': {'overall_median_earnings': 29667,
       'working_not_enrolled': {'overall_count': 41}}},
     'debt': {'staff_grad_plus': {'all': {'eval_inst': {'median': 9500}}}}},
    {'code': '5108',
     'title': 'Allied Health and Medic

352 degrees available with filtered 4 year median. we will use df2 moving forward

In [43]:
df2.isna().any()


code                                                False
title                                               False
unit_id                                             False
distance                                            False
school.type                                         False
school.main_campus                                  False
credential.level                                    False
earnings.4_yr.overall_median_earnings               False
earnings.4_yr.working_not_enrolled.overall_count    False
debt.staff_grad_plus.all.eval_inst.median           False
school.name                                         False
dtype: bool

In [46]:
df2.info()

<class 'pandas.DataFrame'>
MultiIndex: 352 entries, ('Concorde Career Institute-Jacksonville', 'Dental Support Services and Allied Professions.') to ('Indian River State College', 'Business Administration, Management and Operations.')
Data columns (total 9 columns):
 #   Column                                            Non-Null Count  Dtype
---  ------                                            --------------  -----
 0   code                                              352 non-null    str  
 1   unit_id                                           352 non-null    int64
 2   distance                                          352 non-null    int64
 3   school.type                                       352 non-null    str  
 4   school.main_campus                                352 non-null    int64
 5   credential.level                                  352 non-null    int64
 6   earnings.4_yr.overall_median_earnings             352 non-null    int64
 7   earnings.4_yr.working_not_enrolled.

the earnings and debt features should be floats

In [44]:
df2=df2.set_index(["school.name","title"])

In [52]:
df2["debt.staff_grad_plus.all.eval_inst.median"] = df2["debt.staff_grad_plus.all.eval_inst.median"].astype(float)
df2["earnings.4_yr.working_not_enrolled.overall_count"] = df2["earnings.4_yr.working_not_enrolled.overall_count"].astype(float)
df2["earnings.4_yr.overall_median_earnings"] = df2["earnings.4_yr.overall_median_earnings"].astype(float)
df2.head()

code  \
school.name                            title                                                      
Concorde Career Institute-Jacksonville Dental Support Services and Allied Professions.     5106   
                                       Health and Medical Administrative Services.         5107   
                                       Allied Health and Medical Assisting Services.       5108   
                                       Allied Health Diagnostic, Intervention, and Tre...  5109   
                                       Allied Health Diagnostic, Intervention, and Tre...  5109   

                                                                                           unit_id  \
school.name                            title                                                         
Concorde Career Institute-Jacksonville Dental Support Services and Allied Professions.      133845   
                                       Health and Medical Administrative Services.          133845   
                                       Allied Health and Medical Assisting Services.        133845   
                                       Allied Health Diagnostic, Intervention, and Tre...   133845   
                                       Allied Health Diagnostic, Intervention, and Tre...   133845   

                                                                                           distance  \
school.name                            title                                                          
Concorde Career Institute-Jacksonville Dental Support Services and Allied Professions.            1   
                                       Health and Medical Administrative Services.                1   
                                       Allied Health and Medical Assisting Services.              1   
                                       Allied Health Diagnostic, Intervention, and Tre...         1   
                                       Allied Health Diagnostic, Intervention, and Tre...         1   

                                                                                                   school.type  \
school.name                            title                                                                     
Concorde Career Institute-Jacksonville Dental Support Services and Allied Professions.     Private, for-profit   
                                       Health and Medical Administrative Services.         Private, for-profit   
                                       Allied Health and Medical Assisting Services.       Private, for-profit   
                                       Allied Health Diagnostic, Intervention, and Tre...  Private, for-profit   
                                       Allied Health Diagnostic, Intervention, and Tre...  Private, for-profit   

                                                                                           school.main_campus  \
school.name                            title                                                                    
Concorde Career Institute-Jacksonville Dental Support Services and Allied Professions.                      1   
                                       Health and Medical Administrative Services.                          1   
                                       Allied Health and Medical Assisting Services.                        1   
                                       Allied Health Diagnostic, Intervention, and Tre...                   1   
                                       Allied Health Diagnostic, Intervention, and Tre...                   1   

                                                                                           credential.level  \
school.name                            title                                                                  
Concorde Career Institute-Jacksonville Dental Support Services and Allied Professions.                    1   
                                    

In [53]:
df2.info()

<class 'pandas.DataFrame'>
MultiIndex: 352 entries, ('Concorde Career Institute-Jacksonville', 'Dental Support Services and Allied Professions.') to ('Indian River State College', 'Business Administration, Management and Operations.')
Data columns (total 9 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   code                                              352 non-null    str    
 1   unit_id                                           352 non-null    int64  
 2   distance                                          352 non-null    int64  
 3   school.type                                       352 non-null    str    
 4   school.main_campus                                352 non-null    int64  
 5   credential.level                                  352 non-null    int64  
 6   earnings.4_yr.overall_median_earnings             352 non-null    float64
 7   earnings.4_yr.wor

In [55]:
df2.to_csv(r"..\data\raw\scorecard\test.csv")

In [63]:
df.index.get_level_values(0).nunique()

20